insights/parsers/vdsm_log.py

```diff
@@ -29,66 +28,108 @@ class VDSMLog(LogFileOutput):
+        re_message_obj = re.compile(r"\[\w.+\](.*?)\(\w+\:\w+\)$")
+        re_level_obj = re.compile(r"INFO|ERROR|DEBUG|WARN")
 
-        # funcname is in brackets before message, so parse those separately
-        fieldnames = ('thread', 'level', 'timestamp', 'module', 'line', 'logname')
         for line in lines:
-            parts = line.split('::', 6)
-            fields = {'raw_line': line}
-            fields.update({k: v for k, v in zip(fieldnames, parts)})
-            if len(parts) == 7:
-                func, msg = parts[6].split(' ', 1)
-                fields['funcname'] = func[1:-1]  # remove surrounding brackets
-                fields['message'] = msg
-            # Did we get a timestamp in there?
-            if 'timestamp' in fields:
-                # Try to convert the datetime if possible
-                try:
-                    fields['datetime'] = datetime.strptime(
-                        fields['timestamp'][:-4],  # skip milliseconds
-                        self.time_format
-                    )
-                except:
-                    pass
-            yield fields
+            if re_datetime_obj.match(line):
+                # VDSM version 4 log parser
+                fields = dict()
+                thread_and_module = re_thread_and_module.findall(line)
+                timestamp = re_datetime_obj.search(line).group()
+                fields['level'] = re_level_obj.search(line).group()
+                fields['thread'] = thread_and_module[0]
+                fields['logname'] = re_name_obj.findall(line)[0]
+                fields['message'] = re_message_obj.findall(line)[0].strip()
+                fields['module'], fields['lineno'] = thread_and_module[-1].split(':')  # noqa
+                fields['asctime'] = datetime.strptime(timestamp, time_format)
+                yield fields
+            else:
+                # VDSM version 3 log parser
+                #
+                # (psachin): I tried to keep the code unchanged as possible.
+                #
+                # If the line is too short, for some reason, then as many
+                # fields as possible are pulled from the line.
+                fieldnames = ('thread', 'level', 'timestamp', 'module', 'line', 'logname')
+                fields = dict()
+                parts = line.split('::', 6)
+                fields.update({k: v for k, v in zip(fieldnames, filter(None, parts))})
+                if len(parts) == 7:
+                    func, msg = parts[6].split(' ', 1)
+                    fields['message'] = msg
+                # Did we get a timestamp in there?
+                if 'timestamp' in fields:
+                    # Try to convert the datetime if possible
+                    try:
+                        fields['asctime'] = datetime.strptime(
+                            fields['timestamp'], time_format)
+                        del fields['timestamp']
+                    except:
+                        pass
```

Comment:
@psachin would it make sense to store the raw value of `timestamp` in `fields[asctime]` or in `fields[timestamp]` if the conversion fails?  Would that information still be useful?

Reply:
Hi @bfahr 

I decided to use `fields[asctime]`(datetime) compared to `field[timestamp]`(raw) because `get_after()` takes `datetime`. This also makes both version 3 and 4 output same which makes job easy for rule writer. This is the reason I'm deliberately deleting that key-value after conversion. The conversion will fail only if `field[timestamp]` does not comply `time_format` which is rare.

`field[timestamp]` might be useful to show user when exactly the error occurred but it will be more appropriate to convert `datetime` to human readable format. Apart from this I think having an additional `field[timestamp]` is of no use. It is quite easy to include raw time in future if we feel so. 

Let me know your opinion.

Reply:
@psachin that sounds good to me.